### DISCUSSION TITLE
vfox backend

### DISCUSSION THREAD

**Discussion by jdx** [jdx/mise#2051] [General]
I've started the work to onboard vfox as a backend for mise. This will be different than other backends like npm and go and more like asdf which as you might expect is completely different than the rest. vfox plugins will pave the way for windows support but also will give us a more modern plugin interface that will benefit unix users and plugin developers as well.

Note that this is somewhat experimental, vfox support is not guaranteed but I do think it is more likely than not based on my experiments and investigation into this. I've been considering an alternative to asdf since before even starting rtx/mise and I think vfox is the most promising candidate by far. Vfox plugins are in lua which allows the mise CLI to have a lot of control over how plugins execute and making them run cross-platform.

That said, I also see the long-term future of mise being able to support several backends like asdf and vfox. I suspect that when installing something without specifying a backend like `mise install foo@latest`, mise may check the following:

1. core plugin
2. vfox plugin
3. asdf plugin

Reason being is core plugins will offer the most well maintained support for languages that can take advantage of being inside the mise CLI itself. This will always remain a small set of the most popular languages that is reasonable for me to maintain myself alongside other mise contributors. Then we will use vfox plugins which provide windows support and potentially features like progress bars and similar UX components. Then asdf which at least currently offers by far the largest set of plugins for users.

I likely will do this with a custom set of vfox plugins which won't necessarily match the set provided by vfox itself, that way we can onboard each plugin individually once they've had a bit of testing done.

This list will be configurable, so if for some reason you want to always default to asdf plugins over core/vfox you would be able to do that. Of course windows users likely won't be able to use asdf plugins.

It should also be possible at some point to create mise plugins that offer backends like vfox and asdf so you can make your own backends—however that will come last and I don't think that vfox and asdf themselves will likely use that logic. I do think lua will make the most sense for backend plugins.

Here's the milestones as I see them:

- [ ] (partially complete) [vfox.rs](https://github.com/jdx/vfox.rs/issues/4) library which provides a generic rust interface to vfox plugins
- [ ] create vfox backend in mise
- [ ] create per-plugin backend "fallback" logic (e.g.: `mise install foo@latest` will go core -> vfox -> asdf).

**Comment by BoscoDomingo** [comment]
Awesome initiative, thank you for developing such an awesome tool and maintaining it for all of us to benefit from it!!!

**Reply by W1M0R** [reply]
Could `pixi` also be an alternative backend (it is written in rust), since it also has first-class support for Windows: https://pixi.sh/latest/

It supports cross-platform tasks: https://pixi.sh/latest/features/advanced_tasks/#our-task-runner-deno_task_shell

And cross-platform environment variables: https://pixi.sh/latest/reference/project_configuration/#the-activation-table

**Reply by jdx** [reply]
it could be another alternative backend, but the vfox stuff is mostly implemented right now so that's what I'm working on right now at least

**Reply by W1M0R** [reply]
I also discovered [proto](https://github.com/moonrepo/proto) and [aqua](https://github.com/aquaproj/aqua) along the way, and found them very helpful on Windows. Mentioning it here, in case anyone wants to consider those options as mise backends.

**Reply by W1M0R** [reply]
I see mise has a [windows release asset](https://github.com/jdx/mise/releases/download/v2024.7.3/mise-v2024.7.3-win-x64.zip). Are there known limitations running that version in Windows?

**Reply by TCROC** [reply]
I also share @W1M0R's question.  Are those windows releases ready to be used? 👀

**Reply by jdx** [reply]
Not at the time of this writing

**Comment by jdx** [comment]
this is available, it defaults to being on for windows but if you want to use it on unix you can with things like `mise use vfox:version-fox/vfox-nodejs`.
